crates/binjs_io/src/entropy/predict.rs

```diff
@@ -19,49 +20,159 @@ pub struct Entry<'a, K, T> where K: 'a, T: 'a {
+    }
+}
+impl<K> From<Vec<K>> for ByIndex<Symbol, K> {
+    fn from(by_index: Vec<K>) -> Self {
+        ByIndex {
+            phantom: PhantomData,
+            by_index
+        }
+    }
+}
+
 /// A generic predictor, associating a context and a key to a value.
-#[derive(Debug, Default, Serialize, Deserialize)]
+///
+/// For most use cases, you probably want one of the more specialized predictors.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
 pub struct ContextPredict<C, K, T> where C: Eq + Hash + Clone, K: Eq + Hash + Clone {
-    by_context: HashMap<C, HashMap<K, T>>,
+    by_context: HashMap<C, (HashMap<K, T>, ByIndex<T, K>)>,
 }
 impl<C, K, T> ContextPredict<C, K, T> where C: Eq + Hash + Clone, K: Eq + Hash + Clone {
     pub fn new() -> Self {
         Self {
-            by_context: HashMap::new()
+            by_context: HashMap::new(),
         }
     }
+
+    pub fn get<C2: ?Sized>(&self, context: &C2, key: &K) -> Option<&T>
+        where
+            C: std::borrow::Borrow<C2>,
+            C2: Hash + Eq
+    {
+        let (by_key, _) = self.by_context.get(context)?;
+        by_key.get(key)
+    }
+
     pub fn get_mut<C2: ?Sized>(&mut self, context: &C2, key: &K) -> Option<&mut T>
         where
             C: std::borrow::Borrow<C2>,
             C2: Hash + Eq
     {
-        let by_key = self.by_context.get_mut(context)?;
+        let (by_key, _) = self.by_context.get_mut(context)?;
         by_key.get_mut(key)
     }
 
     pub fn entry(&mut self, context: C, key: &K) -> Entry<K, T> {
         let by_key = self.by_context.entry(context)
-            .or_insert_with(|| HashMap::new());
+            .or_insert_with(|| (HashMap::new(), ByIndex::new()));
         Entry {
-            len: by_key.len(),
-            value: by_key.entry(key.clone()),
+            len: by_key.0.len(),
+            value: by_key.0.entry(key.clone()),
         }
     }
 
     /// The number of states in this predictor.
     pub fn len(&self) -> usize {
         self.by_context.values()
-            .map(|map| map.len())
+            .map(|(map, _)| map.len())
             .sum()
     }
-}
 
+    /// All the contexts known to this predictor.
+    ///
+    /// Used mainly for debugging.
+    pub fn contexts(&self) -> impl Iterator<Item=&C> {
+        self.by_context.keys()
+    }
 
-impl<C, K> InstancesToProbabilities for ContextPredict<C, K, usize> where C: Eq + Hash + Clone, K: Eq + Hash + Clone {
+    /// All the values known to this predictor in a given context.
+    ///
+    /// Used mainly for debugging.
+    pub fn keys_at<C2: ?Sized>(&self, context: &C2) -> Option<impl Iterator<Item=&K>>
+        where
+            C: std::borrow::Borrow<C2>,
+            C2: Hash + Eq
+    {
+        Some(self.by_context.get(context)?
+            .0
+            .keys())
+    }
+}
+impl<C, K> ContextPredict<C, K, Symbol> where C: Eq + Hash + Clone, K: Eq + Hash + Clone {
+    /// Get a value by context and index.
```

Comment:
"Get a key by context and value." or "Get a key by context and symbol's index."

Reply:
Note that we don't actually have the `Symbol` when we call this method, as we get the index from the bit-level reader.